File: backend/app/schemas/recipe_import.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


RecipeImportIssueSeverity = Literal["error", "warning"]


class RecipeImportIssue(BaseModel):
    severity: RecipeImportIssueSeverity
    sheet: str
    row: int | None = None
    field: str | None = None
    message: str


class RecipeImportPreviewItem(BaseModel):
    recipe_key: str
    title: str
    description: str
    category: str
    image_url: str | None = None
    default_servings: int | None = None
    cooking_time: int | None = None
    difficulty: str
    source_type: str
    source_url: str | None = None
    status: str
    ingredient_count: int = 0
    step_count: int = 0
    data_note: str | None = None


class RecipeImportPreview(BaseModel):
    filename: str
    file_size_bytes: int = Field(ge=0)
    sheets: list[str] = Field(default_factory=list)
    recipes_total: int = 0
    recipes_importable: int = 0
    recipes_draft: int = 0
    ingredient_rows: int = 0
    step_rows: int = 0
    recipes: list[RecipeImportPreviewItem] = Field(default_factory=list)
    errors: list[RecipeImportIssue] = Field(default_factory=list)
    warnings: list[RecipeImportIssue] = Field(default_factory=list)
    can_import: bool = False
    truncated: bool = False


class RecipeImportResultItem(BaseModel):
    recipe_key: str
    title: str
    action: Literal["created", "updated"]
    ingredient_count: int
    step_count: int


class RecipeImportResult(BaseModel):
    filename: str
    imported_count: int = 0
    created_count: int = 0
    updated_count: int = 0
    items: list[RecipeImportResultItem] = Field(default_factory=list)

File: frontend/types/recipe-import.ts
export type RecipeImportIssue = {
  severity: "error" | "warning";
  sheet: string;
  row: number | null;
  field: string | null;
  message: string;
};

export type RecipeImportPreviewItem = {
  recipe_key: string;
  title: string;
  description: string;
  category: string;
  image_url: string | null;
  default_servings: number | null;
  cooking_time: number | null;
  difficulty: string;
  source_type: string;
  source_url: string | null;
  status: string;
  ingredient_count: number;
  step_count: number;
  data_note: string | null;
};

export type RecipeImportPreview = {
  filename: string;
  file_size_bytes: number;
  sheets: string[];
  recipes_total: number;
  recipes_importable: number;
  recipes_draft: number;
  ingredient_rows: number;
  step_rows: number;
  recipes: RecipeImportPreviewItem[];
  errors: RecipeImportIssue[];
  warnings: RecipeImportIssue[];
  can_import: boolean;
  truncated: boolean;
};

export type RecipeImportResultItem = {
  recipe_key: string;
  title: string;
  action: "created" | "updated";
  ingredient_count: number;
  step_count: number;
};

export type RecipeImportResult = {
  filename: string;
  imported_count: number;
  created_count: number;
  updated_count: number;
  items: RecipeImportResultItem[];
};
